### DISCUSSION TITLE
How to change the C# parameter  type to dynamic

### DISCUSSION THREAD

**Discussion by hyzx86** [ClearFoundry/ClearScript#485] [Q&A]
Is there a way to convert the input type of C# function to ExpandoObject or dynamic?

Just like https://github.com/microsoft/ClearScript/discussions/479

I'm using OrchardCore ，I’m trying to use ClearScript to instead Jint.

Currently, Jint handles the conversion of JS primitive types to C# primitive types
Let's say I have a C# delegate

``` C#
 var engine = new V8ScriptEngine();
            var methods = scriptingManager.GlobalMethodProviders.SelectMany(x => x.GetMethods());
            if (extraMethods != null)
            {
                methods = methods.Concat(extraMethods);
            }
            foreach (var method in methods)
            {
                engine.Script[method.Name]=method.Method(serviceProvider);
            }
engine.Script.test=Func<DateTime,string>((input)=> input.ToString("yyyy-MM-dd") );
            return engine;
```

If it's in Jint, it can directly convert the input argument to a DateTime
But in ClearScript, it is a wrapped object

**Comment by ClearScriptLib** [comment]
Please clarify. Are you asking about changing the source code? Or are you referring to some kind of conversion at runtime? Please say more about what you're trying to do and what isn't working. Thanks!

**Comment by ClearScriptLib** [comment]
Hi @hyzx86,

>If it's in Jint, it can directly convert the input argument to a DateTime

By default, ClearScript prefers not to perform lossy conversions automatically. However, you can do this:

``` C#
var engine = new V8ScriptEngine(V8ScriptEngineFlags.EnableDateTimeConversion);
```

Cheers!

**Reply by hyzx86** [reply]
The e DateTime above is just one example, there are actually many types.

In the following code, methods contains many other types. Including, but not limited to,
 "string, Boolean, number/int, IDictionary< string,object> / ExpandoObject,Array , etc..."
```
 foreach (var method in methods)
 {
   engine.Script[method.Name]=method.Method(serviceProvider);
 }
```

![image](https://user-images.githubusercontent.com/15613121/217029952-caf10a42-2d5d-4304-80d8-33cecb909a78.png)


In my own project, I can set them to object and convert them to the appropriate type, but I developed it based on the OrchardCore's nuget package ~


I wonder if there is a way to make ClearScript vs. NET type easy to convert?
Thanks！

**Reply by ClearScriptLib** [reply]
> string, Boolean, number/int

Lossless conversion of fundamental types is automatic and enabled by default in ClearScript.

> IDictionary< string,object> / ExpandoObject

ClearScript exposes JavaScript objects as [`ScriptObject`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_ScriptObject.htm), which inherits from [`DynamicObject`](https://learn.microsoft.com/en-us/dotnet/api/system.dynamic.dynamicobject) and provides full access. Is that not sufficient?

> Array

JavaScript arrays can have arbitrarily named properties, so a conversion to .NET `Array` is lossy as well as inefficient, especially if the array is large. However, ClearScript detects JavaScript arrays and exposes them as objects implementing [`IList`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ilist). Is that not sufficient?

**Reply by hyzx86** [reply]
Thanks ，I need to try more
